firewalldb/sql_migration.go

```diff
@@ -0,0 +1,486 @@
+				featureBucket := ftBkt.Bucket(ftName)
+				if featureBucket == nil {
+					return fmt.Errorf("feature bucket "+
+						"%s not found", ftName)
+				}
+
+				featurePairs, err := collectKVPairs(
+					featureBucket, true, perm, rule,
+					fn.Some(groupAlias),
+					fn.Some(string(ftName)),
+				)
+				if err != nil {
+					return fmt.Errorf("collecting "+
+						"feature kv pairs failed: %w",
+						err)
+				}
+
+				params = append(params, featurePairs...)
+
+				return nil
+			})
+		})
+		if err != nil {
+			return nil, fmt.Errorf("collecting session kv pairs "+
+				"failed: %w", err)
+		}
+	}
+
+	return params, nil
+}
+
+// collectKVPairs collects all key-value pairs from the given bucket, and
+// returns them as a slice of kvEntry structs. If the errorOnBuckets parameter
+// is set to true, then the function will return an error if the bucket
+// contains any sub-buckets. Note that when the errorOnBuckets parameter is
+// set to false, the function will not collect any key-value pairs from the
+// sub-buckets, and will just ignore them.
+func collectKVPairs(bkt *bbolt.Bucket, errorOnBuckets, perm bool,
+	ruleName string, groupAlias fn.Option[[]byte],
+	featureName fn.Option[string]) ([]*kvEntry, error) {
+
+	var params []*kvEntry
+
+	return params, bkt.ForEach(func(key, value []byte) error {
+		// If the value is nil, then this is a bucket, which we
+		// don't want to process here, as we only want to collect
+		// the key-value pairs, not the buckets. If we should
+		// error on buckets, then we return an error here.
+		if value == nil {
+			if errorOnBuckets {
+				return fmt.Errorf("unexpected bucket %s found "+
+					"in when collecting kv pairs", key)
+			}
+
+			return nil
+		}
+
+		params = append(params, &kvEntry{
+			perm:        perm,
+			ruleName:    ruleName,
+			key:         string(key),
+			featureName: featureName,
+			groupAlias:  groupAlias,
+			value:       value,
+		})
+
+		return nil
+	})
+}
+
+// insertPair inserts a single key-value pair into the SQL database.
+func insertPair(ctx context.Context, tx SQLQueries,
+	entry *kvEntry) (*sqlKvEntry, error) {
+
+	ruleID, err := tx.GetOrInsertRuleID(ctx, entry.ruleName)
+	if err != nil {
+		return nil, err
+	}
+
+	p := sqlc.InsertKVStoreRecordParams{
+		Perm:     entry.perm,
+		RuleID:   ruleID,
+		EntryKey: entry.key,
+		Value:    entry.value,
+	}
+
+	entry.groupAlias.WhenSome(func(alias []byte) {
+		var groupID int64
+		groupID, err = tx.GetSessionIDByAlias(ctx, alias)
+		if err != nil {
+			err = fmt.Errorf("getting group id by alias %x "+
+				"failed: %w", alias, err)
+			return
+		}
+
+		p.GroupID = sqldb.SQLInt64(groupID)
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	entry.featureName.WhenSome(func(feature string) {
+		id, err := tx.GetOrInsertFeatureID(ctx, feature)
+		if err != nil {
+			err = fmt.Errorf("getting/inserting feature id for %s "+
+				"failed: %w", feature, err)
+			return
+		}
+
+		p.FeatureID = sqldb.SQLInt64(id)
+	})
```

Comment:
error not checked!!

Reply:
Ah snap, good catch :pray:! This shows that linters can be really useful for more than just formatting sometimes 😃!